packages/cli/src/scaling/worker-server.ts

```diff
@@ -0,0 +1,137 @@
+import { rawBodyReader, bodyParser } from '@/middlewares';
+import * as ResponseHelper from '@/response-helper';
+import { ScalingService } from '@/scaling/scaling.service';
+
+/**
+ * Responsible for handling HTTP requests sent to a worker.
+ */
+@Service()
+export class WorkerServer {
+	private port: number;
+
+	private server: Server;
+
+	/**
+	 * @doc https://docs.n8n.io/embed/configuration/#credential-overwrites
+	 */
+	private overwritesLoaded = false;
+
+	constructor(
+		private readonly globalConfig: GlobalConfig,
+		private readonly logger: Logger,
+		private readonly scalingService: ScalingService,
+		private readonly credentialsOverwrites: CredentialsOverwrites,
+		private readonly externalHooks: ExternalHooks,
+	) {
+		this.assertWorker();
+
+		const app = express();
+
+		app.disable('x-powered-by');
+
+		this.server = http.createServer(app);
+
+		this.port = this.globalConfig.queue.health.port;
+
+		const overwritesEndpoint = this.globalConfig.credentials.overwrite.endpoint;
+
+		this.server.on('error', (error: NodeJS.ErrnoException) => {
+			if (error.code === 'EADDRINUSE') throw new PortTakenError(this.port);
+		});
+
+		if (this.globalConfig.queue.health.active) {
+			app.get('/healthz', async (req, res) => await this.healthcheck(req, res));
+		}
+
+		if (overwritesEndpoint !== '') {
+			app.post(`/${overwritesEndpoint}`, rawBodyReader, bodyParser, (req, res) =>
+				this.handleOverwrites(req, res),
+			);
+		}
+	}
+
+	async init() {
+		await new Promise<void>((resolve) => this.server.listen(this.port, () => resolve()));
+
+		await this.externalHooks.run('worker.ready');
+
+		this.logger.info(`\nn8n worker server listening on port ${this.port}`);
+	}
+
+	private async healthcheck(_req: express.Request, res: express.Response) {
+		this.logger.debug('[WorkerServer] Health check started');
+
+		try {
+			await Db.getConnection().query('SELECT 1');
+		} catch (value) {
+			this.logger.error('[WorkerServer] No database connection', ensureError(value));
+
+			return ResponseHelper.sendErrorResponse(
+				res,
+				new ServiceUnavailableError('No database connection'),
+			);
+		}
+
+		try {
+			await this.scalingService.pingQueue();
+		} catch (value) {
+			this.logger.error('[WorkerServer] No Redis connection', ensureError(value));
+
+			return ResponseHelper.sendErrorResponse(
+				res,
+				new ServiceUnavailableError('No Redis connection'),
+			);
+		}
+
+		this.logger.debug('[WorkerServer] Health check succeeded');
+
+		ResponseHelper.sendSuccessResponse(res, { status: 'ok' }, true, 200);
+	}
+
+	private handleOverwrites(
+		req: express.Request<{}, {}, ICredentialsOverwrite>,
+		res: express.Response,
+	) {
+		if (this.overwritesLoaded) {
+			ResponseHelper.sendErrorResponse(res, new CredentialsOverwritesSetError());
+			return;
+		}
+
+		if (req.contentType !== 'application/json') {
+			ResponseHelper.sendErrorResponse(res, new NonJsonBodyError());
+			return;
+		}
+
+		this.credentialsOverwrites.setData(req.body);
```

Comment:
Shouldn't we validate that the overwrites are of correct format?

Reply:
The goal of this PR is to extract out this server so that we can add to it more easily. Let's make any changes to existing legacy behavior as a follow-up?

Reply:
Sure 👍